Azure/iotedge
# Rebase Edge on K8s to 1.0.10 branch

This rebases all changes from Edge on K8s public preview branch into latest release/1.0.10 branch. 
Cherry-picked all commits from merge base to latest edge-on-k8s-public-preview branch into this PR, then cleaned it up.
Most C# code is in the "Kubernetes" project, but some changes affect the "Service" and "Core" projects.

Tests:

TEST:
Bring up k8s based on: https://microsoft.github.io/iotedge-k8s-doc/
only using the images from this PR's build.

[k8s] Update Helm charts to allow for correct proxy settings. (#3641)
TEST:
Set ".Values.iotedged.data.no_proxy" on installation.
Confirm "no_proxy" environment var is set for iotedged and agent pods.

[k8s] Add all root CA certs to TLS connections (#3616)
TEST:
Set up Edge runtime with self-signed certificate. Ensure it starts and all modules connect.
 
[k8s] Allow Edge on K8s modules to join HostNetwork. (#3618) 
TEST:
Set "createOptions.NetworkMode=Host" in a module, ensure it starts and connects to EdgeHub.
 
[k8s] Set resource limits and requests for Edge runtime components #3666
TEST:
Start runtime with these Helm charts. Once pods are running, run
kubectl get pod -n <namespace> <pod> -o yaml
For each pod.
Confirm iotedged has resource limit & request set. and qosClass: Guaranteed
Confirm edgagent has resource limit for all containers and qosClass: Guaranteed
Module will have a resource limit for proxy container, and may or may not have resource for module.
 
[k8s public preview] Have EdgeDeploymentOperator report status to Agent #3232
TEST:
1. Set up a module with bad bind mount:
{"image":"mcr.microsoft.com/media/live-video-analytics:1","createOptions":{"hostConfig":{"Binds":["/home/lvaadmin/samples/output:/var/media/","//:/var/lib/azuremediaservices/"],"LogConfig":{"Config":{"max-size":"10m","max-file":"10"},"Type":""}}},"auth":null}
 
2. Add space at the end of the ACR user name in the deployment.  Use a module that attempts to access the ACR.
 
Both setups should not crash agent. Runtime status of device on portal should show a "500" error.
 
[k8s] Discovered some problems with createOptions parsing. (#3743)
TEST:
Set a module with Cmd, EntryPoint, WorkingDir with different capitalization and ensure these are parsed and assigned in pod.

[k8s] Environment Variables can be null or empty. (#3780)
TEST:
In Agent section of Helm charts, set an empty environment variable, ensure iotedged starts and launches Agent with empty variable.

In any module's environment variable section, make an unset environment variable, and one that is all "=", ensure Pod is created with these environment variables present.
